subsys/net/lib/sntp_sock/sntp_sock.c

```diff
@@ -0,0 +1,232 @@
+	}
+
+	if (LVM_GET_LI(pkt->lvm) > SNTP_LI_MAX) {
+		NET_DBG("Unexpected LI: %d", LVM_GET_LI(pkt->lvm));
+		return -EINVAL;
+	}
+
+	if (LVM_GET_MODE(pkt->lvm) != SNTP_MODE_SERVER) {
+		/* For unicast and manycast, server should return 4.
+		 * For broadcast (which is not supported now), server should
+		 * return 5.
+		 */
+		NET_DBG("Unexpected mode: %d", LVM_GET_MODE(pkt->lvm));
+		return -EINVAL;
+	}
+
+	if (pkt->stratum == SNTP_STRATUM_KOD) {
+		NET_DBG("kiss-o'-death stratum");
+		return -EBUSY;
+	}
+
+	if (ntohl(pkt->tx_tm_s) == 0 && ntohl(pkt->tx_tm_f) == 0) {
+		NET_DBG("zero transmit timestamp");
+		return -EINVAL;
+	}
+
+	if (epoch_time) {
+		u32_t ts = ntohl(pkt->tx_tm_s);
+
+		/* Check if most significant bit is set */
+		if (ts & 0x80000000) {
+			/* UTC time is reckoned from 0h 0m 0s UTC
+			 * on 1 January 1900.
+			 */
+			if (ts >= OFFSET_1970_JAN_1) {
+				*epoch_time = ts - OFFSET_1970_JAN_1;
+			} else {
+				return -EINVAL;
+			}
+		} else {
+			/* UTC time is reckoned from 6h 28m 16s UTC
+			 * on 7 February 2036.
+			 */
+			*epoch_time = ts + 0x100000000 - OFFSET_1970_JAN_1;
+		}
+	}
+
+	return 0;
+}
+
+static int sntp_recv_response(struct sntp_ctx *sntp, u32_t timeout)
+{
+	struct sntp_pkt buf = { 0 };
+	u64_t epoch_time = 0;
+	u64_t tmp = 0;
+	int status;
+	int rcvd;
+
+	if (poll(sntp->sock.fds, sntp->sock.nfds, timeout) < 0) {
+		NET_ERR("Error in poll:%d", errno);
+		status = -errno;
+		goto error_exit;
+	}
+
+	rcvd = recv(sntp->sock.fd, (u8_t *)&buf, sizeof(buf), 0);
+	if (rcvd < 0) {
+		status = -errno;
+		goto error_exit;
+	}
+
+	if (rcvd != sizeof(struct sntp_pkt)) {
+		status = -EMSGSIZE;
+		goto error_exit;
+	}
+
+	status = parse_response((u8_t *)&buf, sizeof(buf),
+				sntp->expected_orig_ts, &tmp);
+	if (status == 0) {
+		epoch_time = tmp;
+	}
+
+error_exit:
+	if (sntp->cb) {
+		sntp->cb(sntp, status, epoch_time, sntp->user_data);
+	}
+
+	return status;
+}
+
+static u32_t get_uptime_in_sec(void)
+{
+	u64_t time;
+
+	time = k_uptime_get_32();
+
+	return time / MSEC_PER_SEC;
+}
+
+int sntp_init(struct sntp_ctx *ctx, struct sockaddr *addr, socklen_t addr_len)
+{
+	int ret;
+
+	if (!ctx || !addr) {
+		return -EFAULT;
+	}
+
+	memset(ctx, 0, sizeof(struct sntp_ctx));
+
+	ctx->sock.fd = socket(addr->sa_family, SOCK_DGRAM, IPPROTO_UDP);
+	if (ctx->sock.fd < 0) {
+		NET_ERR("Failed to create UDP socket %d", errno);
+		return -errno;
+	}
+
+	ret = connect(ctx->sock.fd, addr, addr_len);
+	if (ret < 0) {
+		NET_ERR("Cannot connect to UDP remote : %d", errno);
+		return -errno;
+	}
+
+	ctx->sock.fds[ctx->sock.nfds].fd = ctx->sock.fd;
+	ctx->sock.fds[ctx->sock.nfds].events = POLLIN;
+	ctx->sock.nfds++;
+
+	ctx->is_init = true;
+
+	return 0;
+}
+
+int sntp_request(struct sntp_ctx *ctx,
+		 u32_t timeout,
+		 sntp_resp_cb_t callback,
+		 void *user_data)
+{
+	struct sntp_pkt tx_pkt = { 0 };
+	int ret = 0;
+
+	if (!ctx) {
+		return -EFAULT;
+	}
+
+	if (!ctx->is_init) {
+		return -EINVAL;
+	}
+
+	ctx->cb = callback;
```

Comment:
NULL cb should not be allowed, or then user will never get the result of the request.